examples/request-coalescing.ts

```diff
@@ -0,0 +1,431 @@
+} from './utils/loadGen';
+import {
+  getElapsedMillis,
+  logCoalescingStats,
+  logStats,
+} from './utils/loadGenStatisticsCalculator';
+import {createCache, getCacheClient} from './utils/cache';
+
+const cacheKeys: string[] = [];
+
+for (let i = 0; i < 10; i++) {
+  cacheKeys.push(`cacheKey${i}`);
+}
+
+interface GetAndSetOnlyClient {
+  get(cacheName: string, key: string): Promise<CacheGet.Response>;
+  set(
+    cacheName: string,
+    key: string,
+    value: string
+  ): Promise<CacheSet.Response>;
+}
+
+class MomentoClientWrapperWithCoalescing {
+  private readonly momentoClient: CacheClient;
+  private readonly context: RequestCoalescerContext;
+  private readonly getRequestMap: Map<string, Promise<CacheGet.Response>>;
+  private readonly setRequestMap: Map<string, Promise<CacheSet.Response>>;
+
+  constructor(client: CacheClient, context: RequestCoalescerContext) {
+    this.momentoClient = client;
+    this.context = context;
+    this.getRequestMap = new Map<string, Promise<CacheGet.Response>>();
+    this.setRequestMap = new Map<string, Promise<CacheSet.Response>>();
+  }
+
+  get(cacheName: string, key: string): Promise<CacheGet.Response> {
+    if (this.getRequestMap.has(key)) {
+      this.context.numberOfGetRequestsCoalesced++;
+      const mapResponse = this.getRequestMap.get(key);
+      if (mapResponse !== undefined) {
+        return mapResponse;
+      }
+    }
+    const getPromise = this.momentoClient.get(cacheName, key);
+    this.getRequestMap.set(key, getPromise);
+    return getPromise.finally(() => {
+      this.getRequestMap.delete(key);
+    });
+  }
+
+  set(
+    cacheName: string,
+    key: string,
+    value: string
+  ): Promise<CacheSet.Response> {
+    if (this.setRequestMap.has(key)) {
+      this.context.numberOfSetRequestsCoalesced++;
+      const mapResponse = this.setRequestMap.get(key);
+      if (mapResponse !== undefined) {
+        return mapResponse;
+      }
+    }
+    const setPromise = this.momentoClient.set(cacheName, key, value);
+    this.setRequestMap.set(key, setPromise);
+    return setPromise.finally(() => {
+      this.setRequestMap.delete(key);
+    });
+  }
+}
+
+class RequestCoalescerLoadGen {
+  private readonly loggerFactory: MomentoLoggerFactory;
+  private readonly logger: MomentoLogger;
+  private readonly cacheItemTtlSeconds = 60;
+  private readonly options: BasicLoadGenOptions;
+  private readonly delayMillisBetweenRequests: number;
+  private readonly cacheValue: string;
+
+  private readonly cacheName: string = 'js-loadgen';
+
+  constructor(options: BasicLoadGenOptions) {
+    this.loggerFactory = options.loggerFactory;
+    this.logger = this.loggerFactory.getLogger('request-coalescer-load-gen');
+    this.options = options;
+    this.cacheValue = 'x'.repeat(options.cacheItemPayloadBytes);
+    this.delayMillisBetweenRequests =
+      (1000.0 * this.options.numberOfConcurrentRequests) /
+      this.options.maxRequestsPerSecond;
+  }
+
+  async run(): Promise<void> {
```

Comment:
If you made this function take a `GetSetOnlyClient` as an argument then you could get rid of the second copy of it, right?  You might need to move the banner logging statements up into the calling function, as well as the post-run coalescing stats logging, but the rest of this code is identical to the other function, right?